### PR TITLE
Add more precise but flexible typing for statistics, time, and datetime.

### PR DESCRIPTION
**Pull request by Flameeyes**

### CODE REVIEW

File: stdlib/statistics.pyi [hunk 1]
```diff
@@ -32,8 +32,8 @@ if sys.version_info >= (3, 13):
     __all__ += ["kde", "kde_random"]
 
 # Most functions in this module accept homogeneous collections of one of these types
-_Number: TypeAlias = float | Decimal | Fraction
-_NumberT = TypeVar("_NumberT", float, Decimal, Fraction)
+_Number: TypeAlias = Real | Decimal
```
Comment (AlexWaygood) [inline stdlib/statistics.pyi:35]: I'm afraid type checkers don't understand float to be a subtype of numbers.Real

Comment (Flameeyes) [reply stdlib/statistics.pyi:35]: I'm a bit scared now — I know pyre does, and ISTR mypy did as well?

Is there an easy-to-use playground to run the same typing information across all the supported checkers?

Comment (AlexWaygood) [reply stdlib/statistics.pyi:35]: Mypy has never properly supported the ABCs in the numbers module, and I don't believe pyright has either. There's no cross-typechecker playground, but MyPy's playground is at https://mypy-play.net/?mypy=latest&python=3.12 and pyright's is at https://pyright-play.net/

Comment (AlexWaygood) [reply stdlib/statistics.pyi:35]: You can see evidence of this in the new `openlibrary` hit reported by mypy_primer

Comment (Flameeyes) [reply stdlib/statistics.pyi:35]: Okay this is worse than I expected :(

Not only mypy doesn't support it (I *might* be confusing it with pytype, but it's been years, I'm lost), but neither of them seem to even think that a `* 2` maintains the type.

Do you have a suggestion on how to handle this? Just add `Real` without removing `float` ? Replace it with `SupportsFloat` even if it's not quite the right type?

Comment (AlexWaygood) [reply stdlib/statistics.pyi:35]: The numbers ABCs are sort-of fundamentally broken when it comes to static typing, I'm afraid. It's impossible to add proper annotations to many of the methods for those ABCs in numbers.pyi here in typeshed, and so as a result we deliberately don't have the builtin numeric classes inherit from them in builtins.pyi; this is why type checkers do not understand them.

I'll take a look in a bit and see if I can think of what to do here

Comment (AlexWaygood) [reply stdlib/statistics.pyi:35]: Can I ask what the problem you're trying to solve is here? I feel like our current versions of `_Number` and `_NumberT` here are going to work well for the vast majority of uses of the `statistics` functions. Is there a specific false positive mypy emitted on some code that you're trying to fix?

File: stdlib/datetime.pyi [hunk 2]
```diff
@@ -51,7 +51,7 @@ class date:
     resolution: ClassVar[timedelta]
     def __new__(cls, year: SupportsIndex, month: SupportsIndex, day: SupportsIndex) -> Self: ...
     @classmethod
-    def fromtimestamp(cls, timestamp: float, /) -> Self: ...
+    def fromtimestamp(cls, timestamp: SupportsInt, /) -> Self: ...
```
Comment (AlexWaygood) [inline stdlib/datetime.pyi:54]: This isn't correct:

```pycon
>>> class HasInt:
...     def __int__(self): return 42
... 
>>> import datetime as dt
>>> dt.date.fromtimestamp(HasInt())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer
```

File: stdlib/datetime.pyi [hunk 3]
```diff
@@ -259,14 +259,14 @@ class datetime(date):
     # meaning it is only *safe* to pass it as a keyword argument on 3.12+
     if sys.version_info >= (3, 12):
         @classmethod
-        def fromtimestamp(cls, timestamp: float, tz: _TzInfo | None = ...) -> Self: ...
+        def fromtimestamp(cls, timestamp: SupportsInt, tz: _TzInfo | None = ...) -> Self: ...
```
Comment (AlexWaygood) [inline stdlib/datetime.pyi:262]: Nor is this:

```pycon
>>> dt.datetime.fromtimestamp(HasInt())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer
```

File: stdlib/datetime.pyi [hunk 4]
```diff
@@ -259,14 +259,14 @@ class datetime(date):
...
     else:
         @classmethod
-        def fromtimestamp(cls, timestamp: float, /, tz: _TzInfo | None = ...) -> Self: ...
+        def fromtimestamp(cls, timestamp: SupportsInt, /, tz: _TzInfo | None = ...) -> Self: ...
 
     @classmethod
     @deprecated("Use timezone-aware objects to represent datetimes in UTC; e.g. by calling .fromtimestamp(datetime.UTC)")
-    def utcfromtimestamp(cls, t: float, /) -> Self: ...
+    def utcfromtimestamp(cls, t: SupportsInt, /) -> Self: ...
```
Comment (AlexWaygood) [inline stdlib/datetime.pyi:269]: Nor is this:

```pycon
>>> dt.datetime.utcfromtimestamp(HasInt())
<stdin>:1: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer
```

File: stdlib/time.pyi [hunk 5]
```diff
@@ -68,9 +68,9 @@ class struct_time(structseq[Any | int], _TimeTuple):
     def tm_gmtoff(self) -> int: ...
 
 def asctime(time_tuple: _TimeTuple | struct_time = ..., /) -> str: ...
-def ctime(seconds: float | None = None, /) -> str: ...
-def gmtime(seconds: float | None = None, /) -> struct_time: ...
-def localtime(seconds: float | None = None, /) -> struct_time: ...
+def ctime(seconds: SupportsInt | None = None, /) -> str: ...
+def gmtime(seconds: SupportsInt | None = None, /) -> struct_time: ...
+def localtime(seconds: SupportsInt | None = None, /) -> struct_time: ...
```
Comment (AlexWaygood) [inline stdlib/time.pyi:73]: These are all incorrect as well:

```pycon
>>> import time
>>> class HasInt:
...     def __int__(self): return 42
...
>>> time.ctime(HasInt())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer
>>> time.gmtime(HasInt())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer
>>> time.localtime(HasInt())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'HasInt' object cannot be interpreted as an integer

**Review by AlexWaygood — CHANGES_REQUESTED** [review]
The changes to use `SupportsInt` don't look correct either, unfortunately:

File: stdlib/datetime.pyi (hunk 2 above)
Comment (AlexWaygood) [reply stdlib/datetime.pyi:54]: Maybe you meant `SupportsIndex`?

```pycon
>>> class HasIndex:
...     def __index__(self): return 42
... 
>>> dt.date.fromtimestamp(HasIndex())
datetime.date(1970, 1, 1)
```

**Comment by JelleZijlstra** [comment]
Thanks for contributing! I'm closing this PR for now, because it still has unresolved review feedback (and a merge conflict)
after three months of inactivity. If you are still interested, please feel free to open
a new PR (or ping us to reopen this one).

**Pull request closed without merging**